packages/codemirror-blocks/src/ui/PrimitiveList.tsx

```diff
@@ -12,71 +12,71 @@ import { defaultKeyMap } from "../keymap";
 import { useSelector } from "react-redux";
 import { useDrag } from "react-dnd";
 import { copy } from "../copypaste";
+import { err } from "../edits/result";
 
 require("./PrimitiveList.less");
 
 type BasePrimitiveProps = {
   primitive: LanguagePrimitive;
   className: string;
-  onFocus: (primitive: LanguagePrimitive) => void;
-  onKeyDown: (e: React.KeyboardEvent) => void;
+  onFocus: (e: React.FocusEvent) => void;
 };
 
-export const Primitive = (props: BasePrimitiveProps) => {
-  const { primitive, className, onFocus, onKeyDown } = props;
-
-  const focusedNode = useSelector(selectors.getFocusedNode);
-
-  const [_, connectDragSource, connectDragPreview] = useDrag({
-    type: ItemTypes.NODE,
-    item: { content: primitive.name },
-  });
-
-  const handleKeyDown = (e: React.KeyboardEvent) => {
-    switch (defaultKeyMap[CodeMirror.keyName(e)]) {
-      case "Copy": {
-        e.preventDefault();
-        const node = primitive.getASTNode();
-        copy({ focusedNode }, [node]);
-        say("copied " + primitive.toString());
-        primitive.element?.focus(); // restore focus
-        return;
+export const Primitive = React.forwardRef(
+  (props: BasePrimitiveProps, ref: React.MutableRefObject<HTMLSpanElement>) => {
```

Comment:
Better to type this using the generic type parameters of `forwardRef`

```tsx
export const Primitive = React.forwardRef<HTMLSpanElement, BasePrimitiveProps>(
  (props, ref) => {
    ...
  });
```

Reply:
Ah! THAT'S the typing I was looking for. I plumbed through a bunch of SO threads and didn't find something as clean as this. Thanks.